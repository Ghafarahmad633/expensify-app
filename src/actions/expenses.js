import uuid from "uuid";
import database from '../firebase/firebase'
// this is ADD_EXPENSE

export const addExpenses=(expense)=>({
    type:"ADD_EXPENSE",
    expense

});

export const startAddExpense=(expenseData={})=>{
    return (dispatch,getState)=>{
        const uid=getState().auth.uid;
        const {
            desc='',
            note='',
            createdAt=0,
            amount=0
        }=expenseData
        const expense={desc,note,createdAt,amount}
        return database.ref(`users/${uid}/expenses`).push(expense)
            .then((ref)=>{
                dispatch(addExpenses({
                    id:ref.key,
                    ...expense
                }))
            })

    }
}

// this is REMOVE_EXPENSE
export const removeExpense=({id}={})=>({
    type: "REMOVE_EXPENSE",
    id

})
export const startRemoveExpense=({id})=>{
    return (dispatch,getState)=>{
        const uid=getState().auth.uid
        return database.ref(`users/${uid}/expenses/${id}`).remove()
            .then(()=>{
                dispatch(removeExpense({id}))
            })
    }
}
export const removeExpenseSelected=(id)=>({
    type:"DELETE_SELECTED",
    id
})
// This is editExpense
export const editExpense=(id,updates)=>({
    type:"EDIT_EXPENSE",
    id,
    updates

})
export const startEditExpense=(id,updates)=>{
    return (dispatch,getState)=>{
        const uid=getState().auth.uid
       return database.ref(`users/${uid}/expenses/${id}`).update({
            ...updates
        }).then(()=>{
            dispatch(editExpense(id,updates))
       })
    }
}
export const removeExpenseSelctedRange=({ids})=>({
    type:"REMOVE_EXPENSES_SELECTED_RANGE",
    ids

})

export const startRemoveExpenseSelctedRange=({ids})=>{
    return (dispatch,getState)=>{
        const uid=getState().auth.uid
        const objectToIds={}
        ids.forEach((id)=>{
            objectToIds[id]=null
        })
        return database.ref(`users/${uid}/expenses`).update(objectToIds).then(()=>{
            dispatch(removeExpenseSelctedRange({ids}))
        })
    }
}

export const setExpense=(expenses)=>({
    type:"SET_EXPENSES",
    expenses
})
export const startSetExpense=()=>{
    return(dispatch,getState)=>{
        const uid=getState().auth.uid
       return database.ref(`users/${uid}/expenses`).once('value')
            .then((snapshot)=>{
                const expenses=[]
                snapshot.forEach((childSnaphot)=>{
                    expenses.push({
                        id:childSnaphot.key,
                        ...childSnaphot.val()
                    })
                })
                dispatch(setExpense(expenses))
            })

    }
}