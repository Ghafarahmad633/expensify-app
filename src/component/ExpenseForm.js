import React from 'react'
import moment from 'moment'
import {SingleDatePicker} from 'react-dates'

const now=moment();
console.log(now.format('MMM,Do, YYYY'))
class ExpenseForm extends React.Component{
    constructor(props){
        super(props);
        this.state={
            decs: props.expense?props.expense.desc:"",
            note:props.expense?props.expense.note:"",
            amount:props.expense?(props.expense.amount/100).toString():"",
            createdAt:props.expense?moment(props.expense.createdAt):moment(),
            calendarFocused:false,
            error:undefined,
            buttonText:props.buttonText?props.buttonText:"Submit"
        }
    }
    onChangeDecs=(e)=>{
        const decs=e.target.value
        this.setState(()=>({decs}))
    };
    onNoteChange=(e)=>{

        const note=e.target.value
        this.setState(()=>({note}))

    };
    onAmountChange=(e)=>{

        const amount=e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>({amount}))

        }
    };
    onDateChange=(createdAt)=>{

       if(createdAt){
           this.setState(()=>({createdAt}))
       }

    };
    onFocusChange=({focused})=>{
        this.setState(()=>({calendarFocused: focused}))
    };
    onSubmit=(e)=>{
        e.preventDefault()
        if(!this.state.decs||!this.state.amount){
            this.setState(()=>({error: "Please Fill All Fields"}))
            console.log("Some Error")
        }
        else {
            this.setState(()=>({error: ''}))
            this.props.onSubmit(({
                desc:this.state.decs,
                amount:parseFloat(this.state.amount,10)*100,
                createdAt:this.state.createdAt.valueOf(),
                note:this.state.note

            }))
        }
    }
    render(){
        return(
            <div>
                <form onSubmit={this.onSubmit}>
                    <input
                        name='desc'
                        type='text'
                        placeholder='Decritpion'
                        onChange={this.onChangeDecs}
                        value={this.state.decs}
                    />
                    <input
                        type='text'
                        name='amount'
                        placeholder='Amount'
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={(day)=>false}
                    />
                    <textarea
                        name='note'
                        onChange={this.onNoteChange}
                        value={this.state.note}
                        placeholder='add your note'
                    >

                    </textarea>
                    <button type='submit'>{this.state.buttonText}</button>
                </form>
                {this.state.error!==undefined&&<p>Please Fill out the required fields</p>}
            </div>
        )
    }

}
export default ExpenseForm