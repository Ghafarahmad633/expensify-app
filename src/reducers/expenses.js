// Expense Reducer
const expendDefaultState=[];
export default  (state=expendDefaultState,action)=>{
    switch (action.type) {
        case "ADD_EXPENSE":
            return [
                ...state,
                action.expense
            ];
        case "REMOVE_EXPENSE":
            return state.filter(({id,note,desc})=>{
                console.log("Remove "+id)
                return id!==action.id
            });
        case "EDIT_EXPENSE":
            return state.map((expense)=>{
                if(expense.id===action.id){
                    return {
                        ...expense,
                        ...action.updates
                    }
                }
                else{
                    return expense
                }
            });
        case "REMOVE_EXPENSES_SELECTED_RANGE":
            return state.filter((exp)=>{
                return !action.ids.includes(exp.id)
            });
        case 'SET_EXPENSES':
            return action.expenses
        default:
            return state
    }
}