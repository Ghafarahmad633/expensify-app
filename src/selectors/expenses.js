import moment from 'moment'

// Get Visible

export default(expenses,{text,sortBy,startDate,endDate})=>{

    return expenses.filter((exepnse)=>{
        const createdAtMoment=moment(exepnse.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment,'day'):true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment,'day'):true

        const textmatch = exepnse.desc.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textmatch


    }).sort((a,b)=>{
        if(sortBy==='date'){
            return a.createdAt<b.createdAt?1:-1
        }
        else if(sortBy==='amount'){
            return a.amount<b.amount?1:-1
        }
    })

}