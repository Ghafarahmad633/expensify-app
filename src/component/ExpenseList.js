import React from 'react'
import {connect} from 'react-redux'
import ExpenseListItem from '../component/ExpenseListItem'
import selcectExpense from '../selectors/expenses'
import {addItemToDelete,removeItemToDelete} from '../actions/deleteSelected'
export class ExpenseList extends React.Component{
    constructor(props){
        super(props)
    };
    state={
        selectAll:this.props.selectAllExpense===true?'checked':null
    }
    onSelectChange=(e)=>{
        if(e.target.checked){
            this.props.addItemToDelete({id:e.target.value})

        }
        else if(!e.target.checked){
            this.props.removeItemToDelete({id:e.target.value})
        }



    };
    render(){
        return(
            <div className='content-continer'>
                <div className='list-header'>
                    <div className='show-for-mobile'>Expenses</div>
                    <div className='show-for-desktop'>Expense</div>
                    <div className='show-for-desktop'>Amount</div>
                </div>
               <div className='list-item__body'>
                   {
                       this.props.expenses.length===0?(
                           <div className='list-header__message'>No Expenses Found</div>
                       ):(
                           this.props.expenses.map((expenses)=>(
                               <ExpenseListItem
                                   key={expenses.id}
                                   {...expenses}
                                   onSelectChange={this.onSelectChange}
                                   selectAll={this.state.selectAll}
                               />
                           ))
                       )
                   }
               </div>

            </div>
        )
    }
}

const mapStatProps=(state)=>{
    return{
        expenses:selcectExpense(state.expense,state.filters),
        expenseToDelete:state.selectedToDeleted,
        selectAllExpense:state.selectAllExpense

    }
}
const mapDispatchToProps=(dispatch,props)=>{
    return{
        addItemToDelete:(data)=>dispatch(addItemToDelete(data)),
        removeItemToDelete:(data)=>dispatch(removeItemToDelete(data))
    }
}
export default connect(mapStatProps,mapDispatchToProps)(ExpenseList)