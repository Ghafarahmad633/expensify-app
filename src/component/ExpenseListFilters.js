import React from 'react'
import {connect} from 'react-redux'
import {setFiltertext} from "../actions/filters";
import {sortyByDate,sortyByAmount} from '../actions/filters'
import {setStartDate,setEndDate} from '../actions/filters'
import {removeExpenseSelctedRange,startRemoveExpenseSelctedRange} from '../actions/expenses'
import {refereshStateAfterDelete} from '../actions/deleteSelected'
import {DateRangePicker} from 'react-dates'
export class ExpenseListFilters extends React.Component {
    constructor(props){
        super(props)
    }
    state={
        calendarFocused:null,

    }
    onDatesChange=({startDate,endDate})=>{
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)

    };
    onFocusChange=(calendarFocused)=>{
        this.setState(()=>({calendarFocused}))
    };
    deletedAllSelected=()=>{
        this.props.startRemoveExpenseSelctedRange({ids:this.props.selectedToDeleted})
        this.props.refereshStateAfterDelete()


    }

    onTextChange=(e)=>{
        this.props.setFiltertext(e.target.value)
    }
    onSortChange=(e)=>{
        const filter_select=e.target.value
        if(filter_select==='date'){
            this.props.sortyByDate()
        }
        else if(filter_select==='amount'){
            this.props.sortyByAmount()
        }
    }


    render(){
        return(
            <div className='content-continer'>
                <div className='input-group'>
                    <div className='input-group__item'>
                        <input type='text'
                               placeholder='search expenses'
                               className='text-input'
                               disabled={this.props.expenses.length<=0}
                               value={this.props.filters.text} onChange={this.onTextChange}
                        />
                    </div>
                    <div className='input-group__item'>
                        <input type='button' value='Delected Selected' className='button button__form'
                               disabled={this.props.selectedToDeleted.length<=0}
                               onClick={this.deletedAllSelected}
                        />
                    </div>
                    <div className='input-group__item'>
                        <select className='select'
                            disabled={this.props.expenses.length<=0}
                            value={this.props.filters.sortyBy}
                            onChange={this.onSortChange}>
                            <option  value='date'>Date</option>
                            <option value='amount'>Amount</option>
                        </select>
                    </div>
                    <div className='input-group__item'>
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            numberOfMonths={1}
                            isOutsideRange={()=>false}
                            showClearDates={true}
                            disabled={this.props.expenses.length<=0}
                        />
                    </div>
                </div>







            </div>
        )
    }

}

const mapStateProps=(state)=>{
    return{
        filters:state.filters,
        expenses:state.expense,
        selectedToDeleted: state.selectedToDeleted,
        selectAllExpense:state.selectAllExpense
    }
}
const mapDispatchToProps=(dispatch,props)=>{
    return{
        sortyByDate:()=>dispatch(sortyByDate()),
        sortyByAmount:()=>dispatch(sortyByAmount()),
        setFiltertext:(text)=>dispatch(setFiltertext(text)),
        setStartDate:(date)=>dispatch(setStartDate(date)),
        setEndDate:(date)=>dispatch(setEndDate(date)),
        startRemoveExpenseSelctedRange:(data)=>dispatch(startRemoveExpenseSelctedRange(data)),
        refereshStateAfterDelete:()=>dispatch(refereshStateAfterDelete())

    }

}
export default connect(mapStateProps,mapDispatchToProps)(ExpenseListFilters)