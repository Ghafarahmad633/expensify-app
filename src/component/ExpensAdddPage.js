import React from "react";
import ExpenseForm from './ExpenseForm'
import {connect} from 'react-redux'
import {startAddExpense} from '../actions/expenses'
export class ExpensAdddPage extends React.Component{
    onSubmit=(expenses)=>{
        this.props.startAddExpense(expenses)
        this.props.history.push("/")
    }
    render(){
       return(
           <div>
               <div className='page-header'>
                   <div className='content-continer'>
                       <h1 className='page-header__title'>Add Expense</h1>
                   </div>
               </div>

               <div className='content-continer'>
                   <ExpenseForm onSubmit={this.onSubmit}/>
               </div>
           </div>
       )
    }
}
const mapDispatchToProps=(dispatch)=>{
    return{
        startAddExpense:(expenses)=>dispatch(startAddExpense(expenses))
    }
}
export default connect(undefined,mapDispatchToProps)(ExpensAdddPage)