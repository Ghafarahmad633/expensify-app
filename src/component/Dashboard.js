import React from "react";

import ExpenseList from "./ExpenseList";
import ExpenseListFilters from './ExpenseListFilters'
import TotalExpense from './TotalExpense'
class ExpensifyDashboardPage extends React.Component{

    render() {
        return (
            <div>
                <div>
                    <TotalExpense/>
                    <ExpenseListFilters/>
                    <ExpenseList />
                </div>
            </div>
        );
    }
}

export default ExpensifyDashboardPage