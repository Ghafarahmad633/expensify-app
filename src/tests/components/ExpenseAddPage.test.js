import React from 'react'
import {shallow} from 'enzyme'
import {ExpensAdddPage} from '../../component/ExpensAdddPage'
import expenses from '../fixtures/expenses'

let addExpenses,history,wraper;
beforeEach(()=>{
    addExpenses=jest.fn()
     history={push:jest.fn()}
     wraper=shallow(<ExpensAdddPage addExpenses={addExpenses} history={history}/>)
})
test('should render ExpensAdddPage',()=>{

    expect(wraper).toMatchSnapshot()

})
test('should handle onSubmit',()=>{
    wraper.find('ExpenseForm').prop('onSubmit')(expenses[0])
    expect(addExpenses).toHaveBeenLastCalledWith(expenses[0])
    expect(history.push).toHaveBeenLastCalledWith('/')
})