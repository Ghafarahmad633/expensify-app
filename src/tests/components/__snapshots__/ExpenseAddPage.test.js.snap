// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensAdddPage 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-continer"
    >
      <h1
        className="page-header__title"
      >
        Add Expense
      </h1>
    </div>
  </div>
  <div
    className="content-continer"
  >
    <ExpenseForm
      onSubmit={[Function]}
    />
  </div>
</div>
`;
